File: PES-U/beneficiary-dedup-engine/python/models/dedup_result.py
"""
models/dedup_result.py   <->   lib/src/models/dedup_result.dart

The result of scoring one candidate pair.

Verdict is one of: "DUPLICATE", "REVIEW", "CLEAR".
feature_scores carries every individual algorithm score for transparency and
for the in-app warning UX ("matched on: name 0.94, dob 1.0, gps 0.97").

Dart port note: class with final fields; verdict as a String or a Dart enum.
feature_scores -> Map<String,double>. flags -> List<String>.
"""

from dataclasses import dataclass, field
from typing import Dict, List


# Verdict constants — mirror as an enum in Dart.
DUPLICATE = "DUPLICATE"
REVIEW = "REVIEW"
CLEAR = "CLEAR"


@dataclass
class DedupResult:
    id_a: str
    id_b: str
    score: float
    verdict: str
    feature_scores: Dict[str, float] = field(default_factory=dict)
    flags: List[str] = field(default_factory=list)

    def is_duplicate(self) -> bool:
        return self.verdict == DUPLICATE

    def needs_review(self) -> bool:
        return self.verdict == REVIEW

    def is_possible_sibling(self) -> bool:
        return "POSSIBLE_SIBLING" in self.flags

    def top_signals(self, k: int = 3):
        """Highest-scoring features — drives the warning UX explanation."""
        items = sorted(self.feature_scores.items(), key=lambda kv: kv[1], reverse=True)
        return items[:k]

File: PES-U/beneficiary-dedup-engine/dedup_engine/lib/src/models/dedup_result.dart
import 'dedup_config.dart' show Verdict;

/// The result of scoring one candidate pair.
///
/// [featureScores] carries every individual comparison so the host app can
/// explain the match to the user ("matched on: given_name 0.94, dob 1.00").
class DedupResult {
  /// Id of the incoming record.
  final String idA;

  /// Id of the existing record it was compared against.
  final String idB;

  /// Composite score in [0.0, 1.0].
  final double score;

  final Verdict verdict;

  /// Per-comparison scores, keyed by a readable label
  /// (e.g. "given_name:phonetic").
  final Map<String, double> featureScores;

  /// Notes such as "MOBILE_MATCH", "GENDER_MISMATCH", "POSSIBLE_SIBLING".
  final List<String> flags;

  /// The full row of the matched existing record, so the host app can display
  /// it without a second query.
  final Map<String, dynamic> matchedRecord;

  const DedupResult({
    required this.idA,
    required this.idB,
    required this.score,
    required this.verdict,
    this.featureScores = const {},
    this.flags = const [],
    this.matchedRecord = const {},
  });

  bool get isDuplicate => verdict == Verdict.duplicate;
  bool get needsReview => verdict == Verdict.review;
  bool get isClear => verdict == Verdict.clear;

  /// The highest-scoring features — useful for a "why did this match?" line.
  List<MapEntry<String, double>> topSignals([int k = 3]) {
    final entries = featureScores.entries.toList()
      ..sort((a, b) => b.value.compareTo(a.value));
    return entries.take(k).toList();
  }

  @override
  String toString() =>
      'DedupResult($idB, score: ${score.toStringAsFixed(3)}, $verdict)';
}
